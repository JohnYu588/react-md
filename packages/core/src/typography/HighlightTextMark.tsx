import { type ReactElement } from "react";

import { type HighlightTextComponentProps } from "./HighlightText.js";
import { Mark, type MarkProps } from "./Mark.js";

/**
 * @since 6.0.0
 */
export interface HighlightTextMarkProps
  extends HighlightTextComponentProps,
    Omit<MarkProps, "children"> {}

/**
 * This is the default implementation for the `HighlightText` component that
 * just omits the `match` prop from {@link HighlightTextComponentProps} and
 * passes the rest to the {@link Mark}.
 *
 * @since 6.0.0
 */
export function HighlightTextMark(
  props: Readonly<HighlightTextMarkProps>
): ReactElement {
  const { match: _match, ...remaining } = props;
  return <Mark {...remaining} />;
}
