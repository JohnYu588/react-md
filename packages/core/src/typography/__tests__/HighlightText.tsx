import { describe, expect, it } from "@jest/globals";
import { type ReactElement } from "react";

import { render as baseRender, screen } from "../../test-utils/index.js";
import {
  HighlightText,
  type HighlightTextComponentProps,
} from "../HighlightText.js";
import { HighlightTextMark } from "../HighlightTextMark.js";

function MarkWithDataTestId(props: HighlightTextComponentProps): ReactElement {
  return <HighlightTextMark data-testid="mark" {...props} />;
}

const render = (ui: ReactElement) =>
  baseRender(ui, {
    wrapper: function Wrapper({ children }) {
      return <div data-testid="wrapper">{children}</div>;
    },
  });

describe("HighlightText", () => {
  it("should return the children as-is if there is no query, the children are not a string, or the children is an empty string", () => {
    const { rerender } = render(<HighlightText query="">{null}</HighlightText>);

    const wrapper = screen.getByTestId("wrapper");
    expect(wrapper).toBeEmptyDOMElement();

    rerender(<HighlightText query="hello, world!">{null}</HighlightText>);
    expect(wrapper).toBeEmptyDOMElement();

    rerender(
      <HighlightText query="hello, world!">
        <span>Hello, world!</span>
      </HighlightText>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should highlight the text by wrapping in a Mark by default", () => {
    render(
      <HighlightText query="favorite">
        {"I'm Commander Shepard, and this is my favorite store on the Citadel."}
      </HighlightText>
    );
    const wrapper = screen.getByTestId("wrapper");
    expect(wrapper).toHaveTextContent(
      "I'm Commander Shepard, and this is my favorite store on the Citadel."
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should default to find all matches of the query string and wrap in the highlight component", () => {
    const { rerender } = render(
      <HighlightText query="t" highlight={MarkWithDataTestId}>
        {"I'm Commander Shepard, and this is my favorite store on the Citadel."}
      </HighlightText>
    );

    const wrapper = screen.getByTestId("wrapper");
    expect(screen.getAllByTestId("mark")).toHaveLength(5);
    expect(wrapper).toHaveTextContent(
      "I'm Commander Shepard, and this is my favorite store on the Citadel."
    );
    expect(wrapper).toMatchSnapshot();

    rerender(
      <HighlightText query="t" highlight={MarkWithDataTestId} firstMatchOnly>
        {"I'm Commander Shepard, and this is my favorite store on the Citadel."}
      </HighlightText>
    );
    expect(screen.getAllByTestId("mark")).toHaveLength(1);
    expect(wrapper).toHaveTextContent(
      "I'm Commander Shepard, and this is my favorite store on the Citadel."
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should be able to highlight text if the children are an array", () => {
    const show = true;
    const hide = false;
    render(
      <HighlightText query="t" highlight={MarkWithDataTestId}>
        Some prefixed text{show && " with a conditional"}
        {hide && " not valid"}. With a suffix.
      </HighlightText>
    );

    const wrapper = screen.getByTestId("wrapper");
    expect(screen.getAllByTestId("mark")).toHaveLength(5);
    expect(wrapper).toHaveTextContent(
      "Some prefixed text with a conditional. With a suffix."
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should handle diacritics correctly", () => {
    render(
      <HighlightText query="déjà" highlight={MarkWithDataTestId}>
        {"I often experience déjà vu when visiting this place."}
      </HighlightText>
    );

    const wrapper = screen.getByTestId("wrapper");
    expect(() => screen.getByTestId("mark")).not.toThrow();
    expect(wrapper).toHaveTextContent(
      "I often experience déjà vu when visiting this place."
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should ignore whitespace", () => {
    const { rerender } = render(
      <HighlightText query="hello " highlight={MarkWithDataTestId}>
        Hello world
      </HighlightText>
    );

    const wrapper = screen.getByTestId("wrapper");
    const mark = screen.getByTestId("mark");
    expect(mark).toHaveTextContent("Hello");
    expect(wrapper).toHaveTextContent("Hello world");
    expect(wrapper).toMatchSnapshot();

    rerender(
      <HighlightText query=" " highlight={MarkWithDataTestId}>
        Hello world
      </HighlightText>
    );
    expect(() => screen.getByTestId("mark")).toThrow();
    expect(wrapper).toHaveTextContent("Hello world");
    expect(wrapper).toMatchSnapshot();
  });

  it("should work correctly when the match starts at the beginning of the string", () => {
    const loremIpsum =
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. In eget metus ut mi mattis dapibus. Praesent interdum sapien ut posuere convallis. Donec et tristique ex. Aliquam erat volutpat. Donec sit amet dui egestas, tempus quam id, hendrerit risus. Nullam tincidunt quam ut dui aliquet ultricies. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed rhoncus, purus vitae tincidunt placerat, quam diam lobortis magna, et aliquam ante diam id nunc. Cras blandit leo eu nisi elementum, nec gravida ligula pharetra. Etiam molestie luctus orci, vel hendrerit lacus eleifend ut. Nullam placerat dolor ac mi congue, non auctor metus consectetur. Ut pretium mollis vulputate.";
    render(<HighlightText query="lorem ipsum">{loremIpsum}</HighlightText>);

    const wrapper = screen.getByTestId("wrapper");
    expect(wrapper).toHaveTextContent(loremIpsum);
  });
});
