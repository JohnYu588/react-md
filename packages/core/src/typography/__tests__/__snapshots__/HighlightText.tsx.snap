// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightText should be able to highlight text if the children are an array 1`] = `
<div
  data-testid="wrapper"
>
  Some prefixed 
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  ex
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
   w
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    i
  </mark>
  th a cond
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    i
  </mark>
  tional
  . Wi
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  h a suffix.
</div>
`;

exports[`HighlightText should default to find all matches of the query string and wrap in the highlight component 1`] = `
<div
  data-testid="wrapper"
>
  I'm Commander Shepard, and 
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  his is my favori
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  e s
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  ore on 
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  he Ci
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  adel.
</div>
`;

exports[`HighlightText should default to find all matches of the query string and wrap in the highlight component 2`] = `
<div
  data-testid="wrapper"
>
  I'm Commander Shepard, and 
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    t
  </mark>
  his is my favorite store on the Citadel.
</div>
`;

exports[`HighlightText should handle diacritics correctly 1`] = `
<div
  data-testid="wrapper"
>
  I often experience 
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    déjà
  </mark>
   vu when visiting this place.
</div>
`;

exports[`HighlightText should highlight the text by wrapping in a Mark by default 1`] = `
<div
  data-testid="wrapper"
>
  I'm Commander Shepard, and this is my 
  <mark
    class="rmd-mark"
  >
    favorite
  </mark>
   store on the Citadel.
</div>
`;

exports[`HighlightText should ignore whitespace 1`] = `
<div
  data-testid="wrapper"
>
  <mark
    class="rmd-mark"
    data-testid="mark"
  >
    Hello
  </mark>
   world
</div>
`;

exports[`HighlightText should ignore whitespace 2`] = `
<div
  data-testid="wrapper"
>
  Hello world
</div>
`;

exports[`HighlightText should return the children as-is if there is no query, the children are not a string, or the children is an empty string 1`] = `
<div
  data-testid="wrapper"
>
  <span>
    Hello, world!
  </span>
</div>
`;
